### PR TITLE
Fixed How Fields Handle Size

### PR DESCRIPTION
**Pull request by BeksOmega**
## The basics

- [X] I branched from develop
- [X] My pull request is against develop
- [X] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Resolves

#2428 and some collapse/expand problems

### Proposed Changes

1. Makes it so visibility (return a size of 0) is handled inside getSize instead of inside setVisible.
2. Changes (public) updateWidth to (protected) updateSize_.

### Reason for Changes

1. There were problems with collapsing when it was done the previous way. Now all blocks are expanded and collapsed correctly.
2. Makes it less confusing for outside developers who want to create a field with a dynamic height.

Note that the dropdown field's updateWidth function wasn't getting called from anywhere, so it could be safely removed.

### Test Coverage

FieldsHandleSize

I also tested collapsing and collapsing all of the default categories, and it all looks good.

Tested on:
* Desktop Chrome
* Desktop Firefox

Was not able to test on internet explorer or edge because of #2499

### Additional Information

N/A

### CODE REVIEW

File: core/field.js [hunk 1]
```diff
@@ -475,15 +474,24 @@ Blockly.Field.prototype.updateColour = function() {
...
-  this.updateWidth();
+  this.updateSize_();
 };
 
 /**
- * Updates the width of the field. This calls getCachedWidth which won't cache
- * the approximated width on IE/Edge when `getComputedTextLength` fails. Once
- * it eventually does succeed, the result will be cached.
+ * Updates the width of the field. Redirects to updateSize_().
+ * @deprecated May 2019 Use Blockly.Field.getCachedWidth() directly instead.
```
Comment (rachel-fenichel) [inline core/field.js]: Why does this say to use `getCachedWidth` instead of `updateSize_`?

File: core/field_checkbox.js [hunk 2]
```diff
@@ -91,6 +91,14 @@ Blockly.FieldCheckbox.prototype.SERIALIZABLE = true;
  */
 Blockly.FieldCheckbox.prototype.CURSOR = 'default';
 
+/**
+ * Used to tell if the field needs to render. Checkbox fields are statically
```
Comment (rachel-fenichel) [inline core/field_checkbox.js]: "Used to tell if the field needs to be rendered the next time the block is rendered.  Checkbox fields are statically sized, and only need to be rendered at initialization."

^ If this is inaccurate, let's find a more accurate way to say it.

Comment (rachel-fenichel) [reply core/field_checkbox.js]: Similar/the same wording for other places where you're setting isDirty to false.

File: core/field_dropdown.js [hunk 3]
```diff
@@ -510,28 +510,6 @@ Blockly.FieldDropdown.prototype.renderSelectedText_ = function() {
   this.size_.width = Blockly.Field.getCachedWidth(this.textElement_);
 };
 
-/**
- * Updates the width of the field. Overrides field.prototype.updateWidth to
- * deal with image selections on IE and Edge. If the selected item is not an
- * image, or if the browser is not IE / Edge, this simply calls the parent
- * implementation.
- */
-Blockly.FieldDropdown.prototype.updateWidth = function() {
```
Comment (rachel-fenichel) [inline core/field_dropdown.js]: Why did you remove this?

Comment (BeksOmega) [reply core/field_dropdown.js]: It doesn't actually get called anywhere, all of the sizing logic is handled in the image and text render functions.

File: core/field.js (hunk 1 above)
Comment (BeksOmega) [reply core/field.js]: render_ only supports text fields. updateSize_ is only called by the default render_ function. updateSize_ also only supports text fields.

If you have a text field, there is no need to override render_, and no need to call updateSize_.

If you have an image, or mixed field you need to override render_. In this case updateSize_ will not be helpful, because it only supports text fields. But if it is a mixed field, you may still want to know the size of your text, so you should call getCachedWidth().

File: core/field_checkbox.js (hunk 2 above)
Comment (BeksOmega) [reply core/field_checkbox.js]: Fixed

File: core/field.js (hunk 1 above)
Comment (rachel-fenichel) [reply core/field.js]: I see your point, but the redirect still feels wrong to me:
- `updateWidth` and `updateSize` have no return value, but have side effects
- `getCachedWidths` has a return value and no side effects

I suggest:
@deprecated May 2019 Use Blockly.Field.updateSize_() to force an update to the size of the field, or Blockly.Field.getCachedWidth() to check the size of the field..

Comment (BeksOmega) [reply core/field.js]: Sounds good, Fixed.
